payment/watcher.py

```diff
@@ -1,80 +1,158 @@
-        CursorManager.save(cursor)
+        self.enqueuer.enqueue_payment_callback(self.webhook, address, payment)
+        self.remove(address, order_id)
 
-    return cursor
+    async def worker(self):
+        """Poll blockchain and apply callback on watched address. run until stopped."""
+        ledger_seq = await self.get_last_ledger_sequence()
+        retry_count = 0
+        while True:
+            if retry_count == 0:  # Otherwise, stay with the old start_time
+                start_t = time.time()
+            try:
+                addresses = self.watched_addresses.keys()
+                log.debug(f'Watching for ledger {ledger_seq}, try: #{retry_count + 1}')
 
+                try:
+                    txs = await self.bc_manager.get_txs_per_ledger(ledger_seq)
+                    retry_count = 0
+                except ResourceNotFoundError:
+                    log.debug(f'Ledger {ledger_seq} not found')
+                    retry_count += 1
+                    if retry_count >= RETRIES_BEFORE_EXCEPTION:
+                        raise
 
-def on_payment(address, payment):
-    """handle a new payment from an address."""
-    log.info('got payment', address=address, payment=payment)
-    statsd.inc_count('payment_observed',
-                     payment.amount,
-                     tags=['app_id:%s' % payment.app_id,
-                           'address:%s' % address])
+                    log.debug(f'retrying in {WAIT_WHEN_NOT_FOUND} seconds')
+                    await asyncio.sleep(WAIT_WHEN_NOT_FOUND)
+                    continue
 
-    for service_id in Watcher.get_subscribed(address):
-        enqueue_payment_callback(Service.get(service_id), address, payment)
+                if len(txs) != 0:
+                    """
+                    get_txs_per_ledger returns a list of responses from horizon.
+                    Building the Transaction object, which includes decoding the envelops of potentially hundreds of txs 
+                    is cpu intensive and can take a lot of time, (~80ms per 200 txs with 1 op each)
+                    we shouln't block the event loop for that long, so run it in a separate process
+                    """
 
+                    txs = await self.loop.run_in_executor(self.executor, get_kin_payments, txs)
+                    for tx in txs:
+                        try:
+                            _, _, order_id = tx.memo.split('-')
+                        except (AttributeError, ValueError):
+                            continue  # Not one of our orders
 
-def worker(stop_event):
-    """Poll blockchain and apply callback on watched address. run until stopped."""
-    cursor = 0
-    while stop_event is None or not stop_event.is_set():
-        time.sleep(SEC_BETWEEN_RUNS)
-        start_t = time.time()
+                        if (tx.source in addresses and
+                                order_id in self.watched_addresses[tx.source]):
```

Comment:
`not tx.operation.source`?

Reply:
In a SimplifiedTransaction object, tx.source is the source of the payment operation

https://github.com/kinecosystem/kin-sdk-python/blob/v2-master/kin/transactions.py#L57